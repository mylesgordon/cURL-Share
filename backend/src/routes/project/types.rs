use actix_session::SessionGetError;
use actix_web::HttpResponse;
use serde::{Deserialize, Serialize};

use crate::routes::types::UserError;

#[derive(Debug)]
pub enum ProjectError {
    Forbidden(String),
    ProjectDoesNotExistError(String),
    SessionGetError(String),
    SqlxError(sqlx::Error),
    UserError(UserError),
}

impl From<SessionGetError> for ProjectError {
    fn from(e: SessionGetError) -> Self {
        match e {
            // TODO: improve
            _ => ProjectError::SessionGetError(e.to_string()),
        }
    }
}

impl From<sqlx::Error> for ProjectError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => {
                ProjectError::ProjectDoesNotExistError("Project was not found".to_string())
            }
            _ => ProjectError::SqlxError(e),
        }
    }
}

impl From<UserError> for ProjectError {
    fn from(e: UserError) -> Self {
        match e {
            // TODO: improve
            UserError::SessionGetError(e) => ProjectError::SessionGetError(e),
            _ => ProjectError::UserError(e),
        }
    }
}

impl From<ProjectError> for HttpResponse {
    fn from(e: ProjectError) -> Self {
        match e {
            ProjectError::ProjectDoesNotExistError(_) => HttpResponse::NotFound().finish(),
            ProjectError::SessionGetError(_) => HttpResponse::Unauthorized().finish(),
            ProjectError::Forbidden(_) => HttpResponse::Forbidden().finish(),
            _ => HttpResponse::InternalServerError().finish(),
        }
    }
}

#[derive(Serialize)]
pub struct ProjectId {
    pub id: i64,
}

#[derive(Debug, Default, Deserialize)]
pub struct ProjectParams {
    pub search: String,
    pub labels: Vec<String>,
}
