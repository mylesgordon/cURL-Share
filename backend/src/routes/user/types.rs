use actix_web::HttpResponse;
use secrecy::Secret;
use serde::Deserialize;

#[derive(Debug)]
pub enum UserError {
    PasswordHashError(argon2::password_hash::Error),
    SqlxDatabaseError,
    SqlxError(sqlx::Error),
    UserAlreadyExists,
    UserNotFound(sqlx::Error),
}

impl From<argon2::password_hash::Error> for UserError {
    fn from(e: argon2::password_hash::Error) -> Self {
        Self::PasswordHashError(e)
    }
}

impl From<sqlx::Error> for UserError {
    fn from(e: sqlx::Error) -> Self {
        match e {
            sqlx::Error::RowNotFound => Self::UserNotFound(e),
            sqlx::Error::Database(database_err) => {
                if let Some(database_err_code) = database_err.code() {
                    match database_err_code.as_ref() {
                        "2067" => Self::UserAlreadyExists,
                        _ => Self::SqlxDatabaseError,
                    }
                } else {
                    Self::SqlxDatabaseError
                }
            }
            _ => Self::SqlxError(e),
        }
    }
}

impl From<UserError> for HttpResponse {
    fn from(e: UserError) -> Self {
        match e {
            UserError::UserAlreadyExists => HttpResponse::Conflict().into(),
            UserError::UserNotFound(_) => HttpResponse::Unauthorized().into(),
            _ => HttpResponse::InternalServerError().into(),
        }
    }
}

#[derive(Deserialize)]
pub struct UserRequest {
    pub username: String,
    pub password: Secret<String>,
}
