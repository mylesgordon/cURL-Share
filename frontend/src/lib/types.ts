export type Fetch = typeof fetch;

export enum ButtonType {
	Up = -1,
	Down = 1
}

export type CreateProjectResponse = {
	id: number;
};

export type Curl = {
	id: number;
	name: string;
	description: string;
	rawQuery: string;
};

export type CurlGroup = {
	id: number;
	curls: string;
	description: string;
	labels: string;
	name: string;
	project_id: number;
};

export type Project = {
	info: ProjectInfo;
	admins: Array<string>;
	collaborators: Array<string>;
	groups: Array<CurlGroup>;
};

export type ProjectInfo = {
	id: number;
	environments: string;
	description: string;
	name: string;
	visibility: string;
};

export enum Visibility {
	Public = 'Public',
	Private = 'Private'
}

export const VisibilityOptions = [
	{
		name: 'visibility',
		value: 'Public',
		label: 'Public'
	},
	{
		name: 'visibility',
		value: 'Private',
		label: 'Private'
	}
];
