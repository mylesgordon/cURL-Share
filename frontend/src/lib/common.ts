export function splitAndTrim(input: string): Array<string> {
		return input.split(',').map((item) => item.trim());
	}