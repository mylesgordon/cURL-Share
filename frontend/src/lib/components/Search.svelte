<script lang="ts">
	import { Input } from 'agnostic-svelte';
	import type { CurlGroup, ProjectInfo } from '$lib/types';

	export let curlGroupArray: Array<CurlGroup> = [];
	export let projectInfoArray: Array<ProjectInfo> = [];

	const initialCurlGroupArray = [...curlGroupArray];
	const initialProjectInfoArray = [...projectInfoArray];

	let inputValue = '';
	let selectedLabel = '';

	let labels: Array<string> = [
		...new Set(initialCurlGroupArray.map((group) => group.labels.split(',')).flat(1))
	];

	$: curlGroupArray = initialCurlGroupArray.filter((entry) => {
		const nameConstraint = entry.name.toLowerCase().includes(inputValue.toLowerCase());
		const labelConstraint = selectedLabel ? entry.labels.split(',').includes(selectedLabel) : true;
		return nameConstraint && labelConstraint;
	});
	$: projectInfoArray = initialProjectInfoArray.filter((entry) =>
		entry.name.toLowerCase().includes(inputValue.toLowerCase())
	);
</script>

<div class="flex">
	<Input
		bind:value={inputValue}
		placeholder="Search..."
		isLabelHidden={true}
		data-testid="search-input"
	/>

	{#if initialCurlGroupArray.length}
		<select name="labels" id="labels" data-testid="labels-select" bind:value={selectedLabel}>
			<option value="">Choose a label</option>
			{#each labels as label}
				<option value={label}>{label}</option>
			{/each}
		</select>
	{/if}
</div>

<style>
	#labels {
		padding: 6px;
	}
</style>
