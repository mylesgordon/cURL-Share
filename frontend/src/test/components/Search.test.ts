import { type CurlGroup, type ProjectInfo, Visibility } from '$lib/types';
import { fireEvent, render, screen } from '@testing-library/svelte';
import { get, writable } from 'svelte/store';
import { testMatchingSnapshot } from '../common';
import Search from '$lib/components/Search.svelte';
import html from 'svelte-htm';
import userEvent from '@testing-library/user-event';
import type { ComponentProps } from 'svelte';

describe('Search component', () => {
	const curlGroupArray: Array<CurlGroup> = [
		{
			id: 1,
			project_id: 1,
			name: 'Login Journey',
			description: 'cURL group defining a user journey!',
			labels: 'label 1,label 2',
			curls: ''
		},
		{
			id: 2,
			project_id: 1,
			name: 'Another Journey',
			description: 'cURL group defining another journey!',
			labels: 'label 3,label 4',
			curls: ''
		}
	];
	const projectInfoArray: Array<ProjectInfo> = [
		{
			id: 1,
			name: 'Placeholder project',
			description: 'This is a placeholder project',
			visibility: Visibility.Public,
			environments: 'https://cool.com, https://dev.cool.com'
		},
		{
			id: 2,
			name: 'Another project',
			description: 'This is a placeholder project',
			visibility: Visibility.Public,
			environments: 'https://cool.com, https://dev.cool.com'
		},
		{
			id: 3,
			name: 'Final project',
			description: 'This is a placeholder project',
			visibility: Visibility.Public,
			environments: 'https://cool.com, https://dev.cool.com'
		}
	];

	const curlGroupProps: ComponentProps<Search> = { curlGroupArray };
	const projectInfoProps: ComponentProps<Search> = { projectInfoArray };

	testMatchingSnapshot(Search, curlGroupProps);
	testMatchingSnapshot(Search, projectInfoProps);

	it('searches project info correctly', async () => {
		const projectInfo = writable([...projectInfoArray]);
		render(html`<${Search} bind:projectInfoArray=${projectInfo} />`);

		const input = await screen.findByTestId('search-input');
		await userEvent.type(input, 'Another project');

		expect(get(projectInfo)).toStrictEqual([projectInfoArray[1]]);

		await userEvent.clear(input);
		expect(get(projectInfo)).toStrictEqual(projectInfoArray);
	});

	it('searches curl groups correctly', async () => {
		const curlGroups = writable([...curlGroupArray]);
		render(html`<${Search} bind:curlGroupArray=${curlGroups} />`);

		const input = await screen.findByTestId('search-input');
		await userEvent.type(input, 'Another journey');

		expect(get(curlGroups)).toStrictEqual([curlGroupArray[1]]);

		await userEvent.clear(input);
		expect(get(curlGroups)).toStrictEqual(curlGroupArray);

		const labelsDropdown = await screen.getByTestId('labels-select');
		await fireEvent.change(labelsDropdown, { target: { value: 'label 2' } });

		expect(get(curlGroups)).toStrictEqual([curlGroupArray[0]]);
	});
});
