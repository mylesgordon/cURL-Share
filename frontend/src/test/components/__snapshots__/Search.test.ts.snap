// Vitest Snapshot v1

exports[`Search component > should match snapshot 1`] = `
<body>
  <div>
    <div
      class="flex s-MMN3rRE-BmEv"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label screenreader-only s-JZthjkXlZrq3"
          for=""
        >
          
        </label>
         
        <input
          class="input s-JZthjkXlZrq3"
          data-testid="search-input"
          id=""
          placeholder="Search..."
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
      <select
        class="s-MMN3rRE-BmEv"
        data-testid="labels-select"
        id="labels"
        name="labels"
      >
        <option
          class="s-MMN3rRE-BmEv"
          value=""
        >
          Choose a label
        </option>
        <option
          class="s-MMN3rRE-BmEv"
          value="label 1"
        >
          label 1
        </option>
        <option
          class="s-MMN3rRE-BmEv"
          value="label 2"
        >
          label 2
        </option>
        <option
          class="s-MMN3rRE-BmEv"
          value="label 3"
        >
          label 3
        </option>
        <option
          class="s-MMN3rRE-BmEv"
          value="label 4"
        >
          label 4
        </option>
      </select>
    </div>
    <!--&lt;Search&gt;-->
  </div>
</body>
`;

exports[`Search component > should match snapshot 2`] = `
<body>
  <div>
    <div
      class="flex s-MMN3rRE-BmEv"
    >
      <div
        class="w-100 s-JZthjkXlZrq3"
      >
        <label
          class="label screenreader-only s-JZthjkXlZrq3"
          for=""
        >
          
        </label>
         
        <input
          class="input s-JZthjkXlZrq3"
          data-testid="search-input"
          id=""
          placeholder="Search..."
          type="text"
        />
         
      </div>
      <!--&lt;Input&gt;-->
       
    </div>
    <!--&lt;Search&gt;-->
  </div>
</body>
`;
